### DISCUSSION TITLE
Dynamically switching post-processing shaders

### DISCUSSION THREAD

**Discussion by Pietrek14** [bevyengine/bevy#5785] [Q&A] [answered]
In my game I need to apply a post-processing effect when the player uses a power-up. I based my post-processing around the official example from the Bevy repository. Is it possible to switch shaders during runtime / turn them on/off?

**Comment by wilk10** [comment, marked as answer]
Yes, it should be possible.

In the example there's the `post_processing_pass_layer` component that is used to mark the `Camera2dBundle` and `MaterialMesh2dBundle`. 

If you want to apply a different effect, you'll need different materials (just another type different than `PostprocessingMaterial` in the example, that derives `AsBindGroup` and implements `Material2d`, pointing to the correct shader). 

You'll also need multiple entities with different `MaterialMesh2dBundle`s, one for each material (one material per effect).

When it's time to apply a different effect, despawn the entity with the the `MaterialMesh2dBundle` that is currently applied, and spawn a new entity with the new material for the new effect + add the `post_processing_pass_layer` component (or alternatively remove the current `MaterialMesh2dBundle` and insert the new material bundle into the same entity).

I think it should work, but let me know if it doesn't. Good luck!

**Reply by superdump** [reply]
I was going to suggest removing one material component from the mesh entity and adding another instead.

**Reply by wilk10** [reply]
Ah yes of course, I didn't think about that. That should work too

**Reply by Pietrek14** [reply]
Thanks man, works like a charm. I actually had to reset the whole entity cause the material component is a generic, but after a long fight I finally got it to work.
